File: Map_Editor/Map_Editor/Objects/Node.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Map_Editor.Objects
{
    public enum REGION
    { 
        TOP_LEFT    = 1,
        TOP_RIGHT   = 2,
        BOT_LEFT    = 3,
        BOT_RIGHT   = 4
    }

    public class Node
    {
        #region CONSTANT

        public const int SCREEN_WIDTH = 480;
        public const int SCREEN_HEIGHT = 480;

        #endregion


        //node top left
        private Node _Tl;
        public Node Tl
        {
            get { return _Tl; }
            set { _Tl = value; }
        }

        //node top right
        private Node _Tr;
        public Node Tr
        {
            get { return _Tr; }
            set { _Tr = value; }
        }

        //node bot left
        private Node _Bl;
        public Node Bl
        {
            get { return _Bl; }
            set { _Bl = value; }
        }

        //node bot right
        private Node _Br;
        public Node Br
        {
            get { return _Br; }
            set { _Br = value; }
        }

        //id node
        private int _NodeId;
        public int NodeId
        {
            get { return _NodeId; }
            set { _NodeId = value; }
        }

        //region: tl, tr, bl, br
        private REGION _Region;
        public REGION Region
        {
            get { return _Region; }
            set { _Region = value; }
        }

        //boundarybox
        private Box _BoundaryBox;
        public Box BoundaryBox
        {
            get { return _BoundaryBox; }
            set { _BoundaryBox = value; }
        }

        //list objects it contains
        //only leafnode can contain objects
        private List<TileObject> _ListObject;
        public List<TileObject> ListObject
        {
            get { return _ListObject; }
            set { _ListObject = value; }
        }

        /// <summary>
        /// constructor
        /// </summary>
        public Node(int parentNodeId, REGION region, Box boundaryBox)
        {
            _Tl = null;
            _Tr = null;
            _Bl = null;
            _Br = null;

            _Region = region;
            _NodeId = parentNodeId * 10 + (int)_Region;
            _ListObject = new List<TileObject>();
            //node root
            if (parentNodeId == 0)
                _BoundaryBox = boundaryBox;
            else
            {
                switch (_Region)
                { 
                    case REGION.TOP_LEFT:
                        _BoundaryBox = new Box(
                            boundaryBox.X - boundaryBox.Width/4,
                            boundaryBox.Y + boundaryBox.Height/4,
                            boundaryBox.Width/2,
                            boundaryBox.Height/2
                            );
                        break;
                    case REGION.TOP_RIGHT:
                        _BoundaryBox = new Box(
                            boundaryBox.X + boundaryBox.Width / 4,
                            boundaryBox.Y + boundaryBox.Height / 4,
                            boundaryBox.Width / 2,
                            boundaryBox.Height / 2
                            );
                        break;
                    case REGION.BOT_LEFT:
                        _BoundaryBox = new Box(
                            boundaryBox.X - boundaryBox.Width / 4,
                            boundaryBox.Y - boundaryBox.Height / 4,
                            boundaryBox.Width / 2,
                            boundaryBox.Height / 2
                            );
                        break;
                    case REGION.BOT_RIGHT:
                        _BoundaryBox = new Box(
                            boundaryBox.X + boundaryBox.Width / 4,
                            boundaryBox.Y - boundaryBox.Height / 4,
                            boundaryBox.Width / 2,
                            boundaryBox.Height / 2
                            );
                        break;
                }
            }
        }

        /// <summary>
        /// insert node
        /// </summary>
        public static void InsertNode(Node node, TileObject obj)
        {
            //if object doesn't belong to node, then return
            if (!Box.IsIntersect(node.BoundaryBox, obj.MovementRangeBox))
                return;

            //if node covers all screenhight and screenwidth, then divide it into 4 subnodes
            if (node.BoundaryBox.Width >= SCREEN_WIDTH + 6 && node.BoundaryBox.Height >= SCREEN_HEIGHT + 6)
            {
                if (node._Tl == null)
                    node._Tl = new Node(node._NodeId, REGION.TOP_LEFT, node.BoundaryBox);
                if (node._Tr == null)
                    node._Tr = new Node(node._NodeId, REGION.TOP_RIGHT, node.BoundaryBox);
                if (node._Bl == null)
                    node._Bl = new Node(node._NodeId, REGION.BOT_LEFT, node.BoundaryBox);
                if (node._Br == null)
                    node._Br = new Node(node._NodeId, REGION.BOT_RIGHT, node.BoundaryBox);

                //insert each subnode to node
                InsertNode(node._Tl, obj);
                InsertNode(node._Tr, obj);
                InsertNode(node._Bl, obj);
                InsertNode(node._Br, obj);
            }
            //else if node just covers enough screenwidth or screenheight, then add obj to node
            else
                node.ListObject.Add(obj);
        }
    }
}

File: GameMario/GameMario/Node.cpp
#include "Node.h"


Node::Node(int nodeId, int x, int y, int w, int h)
{
	_NodeId = nodeId;
	_X = x;
	_Y = y;
	_Width = w;
	_Height = h;
	_BoundaryBox =  Box(_X - _Width / 2, _Y + _Height / 2, _Width, _Height);

	_Tl = NULL;
	_Tr = NULL;
	_Bl = NULL;
	_Br = NULL;
}

bool Node::IsEmpty()
{
	if (_Tl != NULL || _Tr != NULL || _Bl != NULL || _Br != NULL)
		return false;
	return true;
}

int Node::GetNodeId()
{
	return _NodeId;
}

Box Node::GetBoundaryBox()
{
	return _BoundaryBox;
}

void Node::InsertObject(std::map<int, Node*> mapQuadTree, GameObject* object, Box objectBox)
{
	//if object doesn't belong to node, then return
	if (!AABBCheck(_BoundaryBox, objectBox))
		return;

	//if node covers all screenhight and screenwidth, then divide it into 4 subnodes
	if (_BoundaryBox.fWidth >= SCREEN_WIDTH + 6 && _BoundaryBox.fHeight >= SCREEN_HEIGHT + 6)
	{
		if (_Tl == NULL)
		{
			_Tl = new Node(
				_NodeId * 10 + eNodePosition::eTopLeft,
				_X - _Width / 4,
				_Y + _Height / 4,
				_Width / 2,
				_Height / 2
				);
			mapQuadTree[_Tl->GetNodeId()] = _Tl;
		}
		if (_Tr == NULL)
		{
			_Tr = new Node(
				_NodeId * 10 + eNodePosition::eTopRight,
				_X + _Width / 4,
				_Y + _Height / 4,
				_Width / 2,
				_Height / 2
				);
			mapQuadTree[_Tr->GetNodeId()] = _Tr;
		}
		if (_Bl == NULL)
		{
			_Bl = new Node(
				_NodeId * 10 + eNodePosition::eBotLeft,
				_X - _Width / 4,
				_Y - _Height / 4,
				_Width / 2,
				_Height / 2
				);
			mapQuadTree[_Bl->GetNodeId()] = _Bl;
		}
		if (_Br == NULL)
		{
			_Br = new Node(
				_NodeId * 10 + eNodePosition::eBotRight,
				_X + _Width / 4,
				_Y - _Height / 4,
				_Width / 2,
				_Height / 2
				);
			mapQuadTree[_Br->GetNodeId()] = _Br;
		}

		float left, right, top, bot;
		left = objectBox.fX;
		right = objectBox.fX + objectBox.fWidth;
		top = objectBox.fY;
		bot = objectBox.fY - objectBox.fHeight;
		//if object places in two diagonals, then add it to node
		if ((left <= _X && right >= _X)
			|| (bot <= _Y && top >= _Y))
		{
			_ListObjects.push_back(object);
		}
		else
		{
			//insert each subnode to node
			_Tl->InsertObject(mapQuadTree, object, objectBox);
			_Tr->InsertObject(mapQuadTree, object, objectBox);
			_Bl->InsertObject(mapQuadTree, object, objectBox);
			_Br->InsertObject(mapQuadTree, object, objectBox);
		}
	}
	//else if node just covers enough screenwidth or screenheight, then add obj to node
	else
		_ListObjects.push_back(object);
}

void Node::Release()
{
	while (_ListObjects.size() > 0)
	{
		std::vector<GameObject*>::iterator itr = _ListObjects.begin();
		(*itr)->Release();
		delete (*itr);
		(*itr) = NULL;
		_ListObjects.erase(itr);
	}
}

Node::~Node()
{
}
